monomerdb/localdb/localdb.go

```diff
@@ -0,0 +1,464 @@
+		UpperBound: bucketTxByHeightAndIndex.Key(endHeightBytes),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("new iterator: %v", err)
+	}
+	defer func() {
+		err = wrapCloseErr(err, iter)
+	}()
+	txs := bfttypes.Txs{}
+	for iter.First(); iter.Valid(); iter.Next() {
+		value, err := iter.ValueAndErr()
+		if err != nil {
+			return nil, fmt.Errorf("get value from iterator: %v", err)
+		}
+		tx := make([]byte, len(value))
+		copy(tx, value)
+		txs = append(txs, tx)
+	}
+	return txs, nil
+}
+
+type getter interface {
+	Get([]byte) ([]byte, io.Closer, error)
+}
+
+func headerByHeight(g getter, heightBytes []byte) (_ *monomer.Header, err error) {
+	headerBytes, closer, err := get(g, bucketHeaderByHeight.Key(heightBytes))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = wrapCloseErr(err, closer)
+	}()
+
+	h := new(monomer.Header)
+	if err := cbor.Unmarshal(headerBytes, &h); err != nil {
+		return nil, fmt.Errorf("unmarshal cbor: %v", err)
+	}
+	return h, nil
+}
+
+func headerByLabel(s *pebble.Snapshot, label eth.BlockLabel) (_ *monomer.Header, err error) {
+	hashBytes, closer, err := get(s, bucketHashByLabel.Key([]byte(label)))
+	if err != nil {
+		return nil, fmt.Errorf("get label hash: %w", err)
+	}
+	defer func() {
+		err = wrapCloseErr(err, closer)
+	}()
+	header, err := headerByHash(s, common.Hash(hashBytes))
+	if err != nil {
+		return nil, fmt.Errorf("header by hash: %w", err)
+	}
+	return header, nil
+}
+
+func headerByHash(s *pebble.Snapshot, hash common.Hash) (_ *monomer.Header, err error) {
+	heightBytes, closer, err := get(s, bucketHeightByHash.Key(hash.Bytes()))
+	if err != nil {
+		return nil, fmt.Errorf("get height by hash: %w", err)
+	}
+	defer func() {
+		err = wrapCloseErr(err, closer)
+	}()
+	header, err := headerByHeight(s, heightBytes)
+	if err != nil {
+		return nil, fmt.Errorf("get header by height: %w", err)
+	}
+	return header, nil
+}
+
+func (db *DB) view(cb func(*pebble.Snapshot) error) (err error) {
+	s := db.db.NewSnapshot()
+	defer func() {
+		err = wrapCloseErr(err, s)
+	}()
+	return cb(s)
+}
+
+func (db *DB) updateIndexed(cb func(*pebble.Batch) error) (err error) {
+	b := db.db.NewIndexedBatch()
+	defer func() {
+		err = wrapCloseErr(err, b)
+	}()
+	if err := cb(b); err != nil {
+		return err
+	}
+	if err := b.Commit(pebble.Sync); err != nil {
+		return fmt.Errorf("commit: %v", err)
+	}
+	return nil
+}
+
+func (db *DB) update(cb func(*pebble.Batch) error) (err error) {
+	b := db.db.NewBatch()
+	defer func() {
+		err = wrapCloseErr(err, b)
+	}()
+	if err := cb(b); err != nil {
+		return err
+	}
+	if err := b.Commit(pebble.Sync); err != nil {
+		return fmt.Errorf("commit: %v", err)
+	}
+	return nil
+}
+
+func get(g getter, key []byte) (_ []byte, _ io.Closer, err error) {
+	value, closer, err := g.Get(key)
+	if err != nil {
+		if errors.Is(err, pebble.ErrNotFound) {
+			return nil, nil, monomerdb.ErrNotFound
+		}
+		return nil, nil, fmt.Errorf("%v", err) // Obfuscate the error.
```

Comment:
Is `fmt.Errorf("%v", err)` the same as just returning `err` directly?

Reply:
Slightly different. Returning the err directly would leak the error type information to the caller (so they can do `errors.Is`, etc.). Technically, this would expand our API to include pebble errors. We wrap errors with `%v` to limit the API surface.